### PR TITLE
Add static `of` accessor methods to `ColorScheme` and `TextTheme`

### LINKED ISSUE #72201: Add ColorScheme.of(context)
## Use case
Now that we have `TextTheme.of(context)` I think we should also have `ColorScheme.of(context)`, as I think most of the time `Theme.of(context)` is called, its just to access the `colorScheme` property.

Writing `Theme.of(context).colorScheme` is very long.
Writing `ColorScheme.of(context)` would be faster and also much cleaner

## Proposal

Add `ColorScheme.of(context)`

### PR DESCRIPTION
**Pull request by amal-stack**
The most common use case to lookup a `ThemeData` instance using `Theme.of(context)` is to access either the `ColorScheme`, the `TextTheme`, or both.

Before this change:
```dart
final colors = Theme.of(context).colorScheme;
final textTheme = Theme.of(context).textTheme;
final primaryTextTheme = Theme.of(context).primaryTextTheme;
```
or
```dart
final ThemeData(
      :colorScheme,
      :textTheme,
      :primaryTextTheme,
    ) = Theme.of(context);
```
After this change:
```dart
final colors = ColorScheme.of(context);
final textTheme = TextTheme.of(context);
final primaryTextTheme = TextTheme.primaryOf(context);
```

### Primary Changes
This PR adds static `of` convenience methods to `ColorScheme` and `TextTheme` that delegate to the `ThemeData`'s respective properties. The methods added are:
* `ColorScheme.of(context)` that returns `Theme.of(context).colorScheme`.
* `TextTheme.of(context)` that returns `Theme.of(context).textTheme`.
* `TextTheme.primaryOf(context)` that returns `Theme.of(context).primaryTextTheme`.

### Side-effects
To allow the above changes to function, this PR adds:
* A `theme.dart` import to `color_scheme.dart` to access to `Theme`.
* A `theme.dart` import to `text_theme.dart` to access to `Theme`.
* A `package:flutter/widgets.dart` import to `text_theme.dart` to access `BuildContext`. 
* The above import allowed getting rid of the same `@docImport` from `text_theme.dart`.

### Documentation updates
This PR also updates the following documentation elements:
* Adds docs to the newly added members.
* Updates `TextTheme`'s docs to instruct using `TextTheme.of(context)` instead of `Theme.of(context).textTheme`.
* Updates `Theme.of` to add a "See also" section to `ColorScheme.of` and `TextTheme.of` since these use cases are among the most common ones for `Theme.of(context)`.

Fixes #72201.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by nate-thegrate — COMMENTED** [review]
This looks like a phenomenal change—pulling the color scheme from the theme data is something I do on a regular basis, and the framework does it a lot as well.

But I do have a couple of concerns.

#### Testing
This pull request needs simple tests—otherwise, in the future somebody could make a change that breaks these new methods and it might go unnoticed.

#### Tech Debt
At some point, we might want to convert `Theme.of(context)` so it uses an [`InheritedModel`](https://main-api.flutter.dev/flutter/widgets/InheritedModel-class.html) and can optimize its notifications more effectively. By adding these static methods today, we might incur a "technical debt" in the form of restricting our future options for API design. For instance, if we expand on this concept to cover every aspect of the theme data, it could be more practical to organize them all under the same namespace, like what we do for MediaQuery (i.e. `Theme.colorSchemeOf`, `Theme.textOf`, etc).

<br>

Overall, I believe the APIs introduced in this PR probably would be the ideal long-term solution… but it's probably best to leave a window to explore all the potential options, rather than landing this one quickly.

**Comment by navaronbracke** [comment]
> This looks like a phenomenal change—pulling the color scheme from the theme data is something I do on a regular basis, and the framework does it a lot as well.
> 
> But I do have a couple of concerns.
> 
> #### Testing
> This pull request needs simple tests—otherwise, in the future somebody could make a change that breaks these new methods and it might go unnoticed.
> 
> #### Tech Debt
> At some point, we might want to convert `Theme.of(context)` so it uses an [`InheritedModel`](https://main-api.flutter.dev/flutter/widgets/InheritedModel-class.html) and can optimize its notifications more effectively. By adding these static methods today, we might incur a "technical debt" in the form of restricting our future options for API design. For instance, if we expand on this concept to cover every aspect of the theme data, it could be more practical to organize them all under the same namespace, like what we do for MediaQuery (i.e. `Theme.colorSchemeOf`, `Theme.textOf`, etc).
> 
> 
> Overall, I believe the APIs introduced in this PR probably would be the ideal long-term solution… but it's probably best to leave a window to explore all the potential options, rather than landing this one quickly.

I do agree that converting the `Theme` to an `InheritedModel` is the way to go here. But since the `Theme` is so big in scope, I wonder if we can convert the `Theme` to become an `InheritedModel`, but introduce the aspects gradually, over multiple PR's? To give some context, the aspect class for MediaQuery is private, so perhaps we could do the following? (if at all possible)

1) Provide a private aspect class (similar to how MediaQuery works), with only the color scheme aspect for now
2) Make `Theme` an inherited model, that uses the private aspect class for the Theme
3) Use the new aspect for the `ColorScheme.of(context)` implementation

One problem with this approach is "where do we define the Theme aspect class?", since its usages are probably in different files (i.e. `color_scheme.dart`). Perhaps using a regular class (so not prefixed with `_` in the name) but not exporting it through the material library is a better option?

File: packages/flutter/lib/src/material/text_theme.dart
Comment (amal-stack) [reply packages/flutter/lib/src/material/text_theme.dart]: Qualified both the methods with their class names. :)

File: packages/flutter/lib/src/material/theme.dart
Comment (amal-stack) [reply packages/flutter/lib/src/material/theme.dart]: Done :)

**Comment by nate-thegrate** [comment]
> Perhaps using a regular class (so not prefixed with `_` in the name) but not exporting it through the material library is a better option?

I don't think this has been done in the framework before, but I'm a fan!
I'm also toying with the idea of giving public access to the class…

```dart
// this PR
final colors = ColorScheme.of(context);
return FilledButton(
  style: FilledButton.styleFrom(
    background: colors.tertiaryContainer,
    foreground: colors.onTertiaryContainer,
  ),
  // ...
);


// eventually:
return FilledButton(
  style: ButtonStyle.aspect(
    surface: const MaterialSurface.tertiaryContainer(),
  ),
  // ...
);
```

If we had a public API that essentially just contained instructions for how to parse & apply the theme data, it could potentially lead to significant performance gains, as dependency-based rebuilds are pushed into the leaf widgets.
(I have a [WIP design doc](https://docs.google.com/document/d/1DYZ3aeOLRqdRj_232y-QhMNzA54bRQwvzPVtQh0MFQ4/edit?usp=sharing) for this, though progress is on hold until I finish up some stuff in the animation library.)

**Review by nate-thegrate — APPROVED** [review]
@amal-stack thanks for the updates and for the thorough explanation; based on what you've shared, I feel good about landing this change. We'll be able to merge once we have 2 approvals.

**Review by QuncCccccc — APPROVED** [review]
LGTM! Thanks for the improvement!

**Comment by justinmc** [comment]
Do we have an open issue anywhere about the Theme as InheritedModel idea? There's a lot of good discussion here that hopefully we can build on.

**Comment by nate-thegrate** [comment]
Yes indeed!

https://github.com/flutter/flutter/issues/154567

**Pull request merged**
